### PR TITLE
Protect kubernetes key pods from eviction

### PR DESCRIPTION
**Pull request by corest**
Towards: https://github.com/giantswarm/giantswarm/issues/2634

- [x] assign k8s cluster components pods `Guaranteed` qos
  - [x] api-server
  - [x] scheduler
  - [x] controller-manager
- [x] assign critical pods `Guaranteed` qos
  - [x] kube-proxy ds
  - [x] calico-node
  - [x] kube-controller
- [x] create priority classes
  - [x] high priority for cluster components
  - [x] high-1 for critical pods
  - [x] default for others
- [x] apply priority classes for pods

### CODE REVIEW

**Comment by corest** [comment]
As a first step of protecting critical pods - we need (or maybe don't?) to give k8s components pods `Guaranteed` qos. This requires settings resources/limits. But those components will require more resources in bigger clusters.

Here is some data regarding resources usage, grabbed from k8s repo issues:

**100 nodes**
```
99 percentile:
container                                                                 cpu(cores) memory(MB)
"kube-apiserver-e2e-test-gmarek-master/kube-apiserver"                    1.233      895.38
"kube-controller-manager-e2e-test-gmarek-master/kube-controller-manager"  0.345      319.96
"kube-scheduler-e2e-test-gmarek-master/kube-scheduler"                    0.232      48.28
```

**300 nodes**
```
99 percentile:
container                                                                 cpu(cores) memory(MB)
"kube-apiserver-e2e-test-gmarek-master/kube-apiserver"                    2.103      1726.28
"kube-controller-manager-e2e-test-gmarek-master/kube-controller-manager"  0.517      916.42
"kube-scheduler-e2e-test-gmarek-master/kube-scheduler"                    1.041      107.36
```

**500 nodes**
```
99 percentile:
container                                                                 cpu(cores) memory(MB)
"kube-apiserver-e2e-test-gmarek-master/kube-apiserver"                    2.998      2439.88
"kube-controller-manager-e2e-test-gmarek-master/kube-controller-manager"  0.633      1624.04
"kube-scheduler-e2e-test-gmarek-master/kube-scheduler"                    1.576      167.41
```

With current limits, we are able to process ~60nodes. I didn't see more in our guest clusters. But even if there will be more - are we going to distribute that load? Or vertical scale one master? Do we have any strategies for those cases? 

cc @puja108 @pipo02mix @calvix @r7vme

**Comment by puja108** [comment]
We have 2 stories in the roadmap for that, 1 is vertical scaling of masters, 1 is multi-master, both can help with scalability of master. 60 nodes are ok for now I would say. Especially as that limit shouldn't be a super hard limit I suppose?

**Comment by corest** [comment]
Well, it is hard limit. Container is now allowed to get more. But from eviction policy perspective - Guaranteed pods will be the last in order to be evicted. Also with proper priorities those pods will be the first in order for rescheduling after cleaning pods with less priority. It is just the thing to consider in future - with such limits and vertical scaling we will have timeouts on api/scheduler etcd rescheduling

File: v_3_2_1/master_template.go [hunk 1]
```diff
@@ -277,6 +275,10 @@ write_files:
               resources:
                 requests:
                   cpu: 250m
+		  memory: 100Mi
```
Comment (calvix) [inline v_3_2_1/master_template.go]: this look like bad indentation

File: v_3_2_1/master_template.go [hunk 2]
```diff
@@ -418,6 +417,9 @@ write_files:
                 requests:
                   cpu: 30m
                   memory: 90Mi
+		limits:
```
Comment (calvix) [inline v_3_2_1/master_template.go]: this look like bad indentation

File: v_3_2_1/master_template.go [hunk 3]
```diff
@@ -377,11 +377,10 @@ write_files:
           - key: node-role.kubernetes.io/master
             operator: Exists
             effect: NoSchedule
-          - key: CriticalAddonsOnly
-            operator: Exists
           # The controllers must run in the host network namespace so that
           # it isn't governed by policy that would prevent it from working.
           hostNetwork: true
+	  priorityClassName: critical-pods
```
Comment (calvix) [inline v_3_2_1/master_template.go]: this look like bad indentation

File: v_3_2_1/master_template.go [hunk 4]
```diff
@@ -515,11 +517,10 @@ write_files:
             k8s-app: coredns
         spec:
           serviceAccountName: coredns
+	  priorityClassName: important-pods
```
Comment (calvix) [inline v_3_2_1/master_template.go]: this look like bad indentation

File: v_3_2_1/master_template.go [hunk 5]
```diff
@@ -861,6 +858,9 @@ write_files:
                 requests:
                   memory: "80Mi"
                   cpu: "75m"
+		limits:
```
Comment (calvix) [inline v_3_2_1/master_template.go]: this look like bad indentation

File: v_3_2_1/master_template.go [hunk 6]
```diff
@@ -829,16 +829,13 @@ write_files:
...
-            scheduler.alpha.kubernetes.io/critical-pod: ''
         spec:
           tolerations:
           - key: node-role.kubernetes.io/master
             operator: Exists
             effect: NoSchedule
-          - key: CriticalAddonsOnly
-            operator: Exists
           hostNetwork: true
+	  priorityClassName: critical-pods
```
Comment (calvix) [inline v_3_2_1/master_template.go]: this look like bad indentation

**Comment by corest** [comment]
Thx, will check. My vim doesn't show those bad indentations

File: v_3_2_1/master_template.go (hunk 1 above)
Comment (corest) [reply v_3_2_1/master_template.go]: Not sure how github creates diff, but this is not valid the last commits

**Pull request merged**